### DISCUSSION TITLE
Audio Player Support in PrairieLearn for Text-to-Speech Feedback

### DISCUSSION THREAD

**Discussion by aalrabah** [PrairieLearn/PrairieLearn#10619] [Dev]
Hi PrairieLearn Team,

I'm exploring the use of OpenAI’s text-to-speech model to let students listen to the output the autograder generates.

Does PrairieLearn support embedding an audio player, or is there a way to implement this feature? Any guidance or workarounds would be appreciated.

**Comment by nwalters512** [comment]
We don't specifically have any first-party support for audio players, but as questions are HTML, we don't need to; you can drop in any audio player you like.

You can include arbitrary information in the JSON produced by an external grader (up to a 1MB limit). One approach would be to include a base64-encoded MP3 in the results and then provide that as the source for an audio player on the client.

**Reply by aalrabah** [reply]
I'm having some trouble implementing the audio player in PrairieLearn. I tried the approach you suggested: generating a base64-encoded MP3 file in the external grader and including it in the JSON results. I also added code to pass this encoded audio to the question.html

However, I’m stuck on how to correctly set up the audio player on the client side to utilize this encoded audio. Could you provide more specific guidance or an example?

Thank you for your help!

**Reply by jonatanschroeder** [reply]
Try adding to the appropriate panel (assuming the base64 content of the file is in `data["feedback"]["mp3base64data"]`):
```html
<object type="audio/mpeg" data="data:audio/mpeg;base64,{{feedback.mp3base64data}}">Fallback text</object>
```
